File: csharp/food-chain/FoodChain.cs
using System.Text;

public static class FoodChain
{
    private static readonly string[] Animals = [
        "fly",
        "spider",
        "bird",
        "cat",
        "dog",
        "goat",
        "cow",
        "horse"
    ];

    private static readonly string?[] Remarks =
    [
        null,
        "It wriggled and jiggled and tickled inside her.",
        "How absurd to swallow a bird!",
        "Imagine that, to swallow a cat!",
        "What a hog, to swallow a dog!",
        "Just opened her throat and swallowed a goat!",
        "I don't know how she swallowed a cow!",
        "She's dead, of course!"
    ];

    public static string Recite(int verseNumber) => BuildVerse(verseNumber - 1);

    public static string Recite(int startVerse, int endVerse)
    {
        var verses = new List<string>();

        for (int i = startVerse; i <= endVerse; i++)
        {
            verses.Add(Recite(i));
        }

        return string.Join("\n\n", verses);
    }

    private static string BuildVerse(int index)
    {
        var sb = new StringBuilder();

        sb.AppendLine($"I know an old lady who swallowed a {Animals[index]}.");

        if (Remarks[index] != null)
        {
            sb.AppendLine(Remarks[index]);
        }

        if (Animals[index] == "horse")
        {
            return sb.ToString().TrimEnd();
        }

        for (int i = index; i > 0; i--)
        {
            var prey = Animals[i - 1];

            if (prey == "spider")
            {
                prey += " that wriggled and jiggled and tickled inside her";
            }

            sb.AppendLine($"She swallowed the {Animals[i]} to catch the {prey}.");
        }

        sb.AppendLine("I don't know why she swallowed the fly. Perhaps she'll die.");

        return sb.ToString().TrimEnd();
    }
}

File: cpp/food-chain/food_chain.cpp
#include "food_chain.h"
#include <vector>

namespace food_chain
{

    std::string verse(int number)
    {
        const std::vector<std::string> animals{
            "fly",
            "spider",
            "bird",
            "cat",
            "dog",
            "goat",
            "cow",
            "horse"};

        const std::vector<std::string> extra{
            "",
            "It wriggled and jiggled and tickled inside her.\n",
            "How absurd to swallow a bird!\n",
            "Imagine that, to swallow a cat!\n",
            "What a hog, to swallow a dog!\n",
            "Just opened her throat and swallowed a goat!\n",
            "I don't know how she swallowed a cow!\n",
            "She's dead, of course!\n"};

        int index = number - 1;
        std::string result;

        result += "I know an old lady who swallowed a " + animals[index] + ".\n";

        if (animals[index] == "horse")
        {
            result += extra[index];
            return result;
        }

        result += extra[index];

        for (int i = index; i > 0; --i)
        {
            result += "She swallowed the " + animals[i] + " to catch the " + animals[i - 1];

            if (animals[i - 1] == "spider")
            {
                result += " that wriggled and jiggled and tickled inside her";
            }

            result += ".\n";
        }

        result += "I don't know why she swallowed the fly. Perhaps she'll die.\n";

        return result;
    }

    std::string verses(int start, int end)
    {
        std::string result;

        for (int i = start; i <= end; ++i)
        {
            result += verse(i);
            result += "\n";
        }

        return result;
    }

    std::string sing()
    {
        return verses(1, 8);
    }

} // namespace food_chain
